File: Semestrul 4/MPP/lab05/java/src/main/java/mpp/repository/abstracts/AbstractDBRepository.java
package mpp.repository.abstracts;

import mpp.domain.Entity;
import mpp.repository.interfaces.Repository;
import mpp.utils.DatabaseConfig;

import java.io.IOException;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Optional;

public abstract class AbstractDBRepository<ID, E extends Entity<ID>> implements Repository<ID, E> {

    protected final String tableName;
    protected final String primaryKey;
    protected final Connection connection;


    public AbstractDBRepository(String tableName, String primaryKey) {
        this.tableName = tableName;
        this.primaryKey = primaryKey;
        try{
            connection = getConnection();
        } catch (Exception e) {
//            log.error(e);
            throw new RuntimeException();
        }
    }

    protected abstract String getInsertParameters();

    protected abstract String getUpdateParameters();

    protected abstract int getUpdateParameterCount();

    /**
     * Creates an entity from a DB set of data
     *
     * @param data - the data to be converted
     * @return the entity
     */
    protected abstract E createEntity(ResultSet data) throws SQLException;

    /**
     * Updates the data of a db row
     *
     * @param entity - the entity to be converted
     */
    protected abstract void saveEntity(PreparedStatement preparedStatement, E entity) throws SQLException;


    @Override
    public Optional<E> save(E entity) {
        if (findOne(entity.getId()).isPresent())
            return Optional.of(entity);
        String sql = String.format("INSERT INTO %s VALUES %s", tableName, getInsertParameters());
        try {
            PreparedStatement statement = connection.prepareStatement(sql);
            saveEntity(statement, entity);

            statement.executeUpdate();
            return Optional.empty();
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public Optional<E> delete(ID id) {
        var entity = findOne(id);
        if (entity.isEmpty())
            return Optional.empty();

        String sql = String.format("DELETE FROM %s WHERE " + primaryKey +" = ?", tableName);
        try {
            PreparedStatement statement = connection.prepareStatement(sql);

            statement.setObject(1, id);
            statement.executeUpdate();

            return entity;
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public Optional<E> update(E entity) {
        String sql = String.format("UPDATE %s SET %s WHERE %s = ?",
                tableName,
                getUpdateParameters(),
                primaryKey);
        try {
            PreparedStatement statement = connection.prepareStatement(sql);
            saveEntity(statement, entity);
            statement.setObject(getUpdateParameterCount() + 1, entity.getId());

            int rowsAffected = statement.executeUpdate();
            return rowsAffected > 0 ? Optional.of(entity) : Optional.empty();
        } catch (SQLException er) {
            throw new RuntimeException(er);
        }
    }


    @Override
    public Optional<E> findOne(ID id) {
        String sql = String.format("SELECT * FROM %s WHERE " + primaryKey + " = ?", tableName);

        try {
            PreparedStatement preparedStatement = connection.prepareStatement(sql);
            preparedStatement.setObject(1, id);

            ResultSet result = preparedStatement.executeQuery();
            if (result.next()) {
                return Optional.of(createEntity(result));
            }
            return Optional.empty();
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public Iterable<E> findAll() {
        List<E> entities = new ArrayList<>();
        String sql = String.format("SELECT * FROM %s", tableName);

        try {
            PreparedStatement statement = connection.prepareStatement(sql);
            ResultSet result = statement.executeQuery();

            while (result.next()) {
                entities.add(createEntity(result));
            }
            return entities;
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    public int size() {
        return ((Collection<?>) findAll()).size();
    }

    protected Connection getConnection() throws SQLException, IOException {
        return DatabaseConfig.getConnection();
    }

    protected Iterable<E> createList(ResultSet resultSet) throws SQLException {
        try {
            List<E> list = new ArrayList<>();
            while (resultSet.next()) {
                E entity = createEntity(resultSet);
                list.add(entity);
            }
            return list;
        } catch (SQLException e) {
            throw new SQLException("Error creating list from ResultSet: " + e.getMessage());
        }
    }
}

File: Semestrul 4/SDI/lab04/cs/Repository/Abstracts/AbstractDBRepository.cs
using cs.Domain;
using cs.Repository.Interfaces;
using cs.Utils;
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.SQLite;
using System.Linq;

namespace cs.Repository.Abstracts;

public abstract class AbstractDBRepository<TId, TEntity> : IRepository<TId, TEntity> where TEntity : IEntity<TId>
{
    protected readonly string _tableName;
    protected readonly string _primaryKey;
    protected readonly SQLiteConnection _connection;

    protected AbstractDBRepository(string tableName, string primaryKey)
    {
        _tableName = tableName;
        _primaryKey = primaryKey;
        try
        {
            _connection = GetConnection();
        }
        catch (Exception ex)
        {
            // Logger.LogError(ex);
            throw new ApplicationException("Failed to establish database connection", ex);
        }
    }

    protected abstract string GetInsertParameters();

    protected abstract string GetUpdateParameters();

    protected abstract int GetUpdateParameterCount();

    /// <summary>
    /// Creates an entity from a database reader
    /// </summary>
    /// <param name="reader">The data reader containing entity data</param>
    /// <returns>The entity</returns>
    protected abstract TEntity CreateEntity(IDataReader reader);

    /// <summary>
    /// Sets parameters for the database command
    /// </summary>
    /// <param name="command">The SQLite command</param>
    /// <param name="entity">The entity to save</param>
    protected abstract void SetParameters(SQLiteCommand command, TEntity entity);

    public virtual TEntity Save(TEntity entity)
    {
        TEntity existingEntity = FindOne(entity.GetId());
        if (existingEntity != null)
            return entity;

        string sql = $"INSERT INTO {_tableName} VALUES {GetInsertParameters()}";
        try
        {
            using var command = new SQLiteCommand(sql, _connection);
            SetParameters(command, entity);

            command.ExecuteNonQuery();
            return default;
        }
        catch (Exception ex)
        {
            throw new ApplicationException($"Failed to save entity to {_tableName}", ex);
        }
    }

    public virtual TEntity Delete(TId id)
    {
        TEntity entity = FindOne(id);
        if (entity == null)
            return default;

        string sql = $"DELETE FROM {_tableName} WHERE {_primaryKey} = @id";
        try
        {
            using var command = new SQLiteCommand(sql, _connection);
            command.Parameters.AddWithValue("@id", id);
            command.ExecuteNonQuery();

            return entity;
        }
        catch (Exception ex)
        {
            throw new ApplicationException($"Failed to delete entity from {_tableName}", ex);
        }
    }

    public virtual TEntity Update(TEntity entity)
    {
        string sql = $"UPDATE {_tableName} SET {GetUpdateParameters()} WHERE {_primaryKey} = @id";
        try
        {
            using var command = new SQLiteCommand(sql, _connection);
            SetParameters(command, entity);
            command.Parameters.AddWithValue("@id", entity.GetId());

            int rowsAffected = command.ExecuteNonQuery();
            return rowsAffected > 0 ? entity : default;
        }
        catch (Exception ex)
        {
            throw new ApplicationException($"Failed to update entity in {_tableName}", ex);
        }
    }

    public virtual TEntity FindOne(TId id)
    {
        string sql = $"SELECT * FROM {_tableName} WHERE {_primaryKey} = @id";

        try
        {
            using var command = new SQLiteCommand(sql, _connection);
            command.Parameters.AddWithValue("@id", id);

            using var reader = command.ExecuteReader();
            if (reader.Read())
            {
                return CreateEntity(reader);
            }
            return default;
        }
        catch (Exception ex)
        {
            throw new ApplicationException($"Failed to find entity in {_tableName}", ex);
        }
    }

    public virtual IEnumerable<TEntity> FindAll()
    {
        List<TEntity> entities = new();
        string sql = $"SELECT * FROM {_tableName}";

        try
        {
            using var command = new SQLiteCommand(sql, _connection);
            using var reader = command.ExecuteReader();

            while (reader.Read())
            {
                entities.Add(CreateEntity(reader));
            }
            return entities;
        }
        catch (Exception ex)
        {
            throw new ApplicationException($"Failed to retrieve entities from {_tableName}", ex);
        }
    }

    public int Count()
    {
        return FindAll().Count();
    }

    protected SQLiteConnection GetConnection()
    {
        return DatabaseConfig.GetConnection();
    }

    protected IEnumerable<TEntity> CreateList(IDataReader reader)
    {
        try
        {
            List<TEntity> list = new();
            while (reader.Read())
            {
                TEntity entity = CreateEntity(reader);
                list.Add(entity);
            }
            return list;
        }
        catch (Exception ex)
        {
            throw new DataException($"Error creating list from data reader: {ex.Message}", ex);
        }
    }
}